spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcWriter.java

```diff
@@ -20,399 +20,121 @@
-      } else {
-        output.isNull[rowId] = false;
-        InternalRow value = data.getStruct(column, children.length);
-        StructColumnVector cv = (StructColumnVector) output;
-        for (int c = 0; c < children.length; ++c) {
-          children[c].addValue(rowId, c, value, cv.fields[c]);
-        }
-      }
-    }
-  }
-
-  static class ListConverter implements Converter {
-    private final Converter children;
-
-    ListConverter(TypeDescription schema) {
-      children = buildConverter(schema.getChildren().get(0));
+    public SparkOrcValueWriter map(Types.MapType iMap, TypeDescription map,
+                                   SparkOrcValueWriter key, SparkOrcValueWriter value) {
+      return SparkOrcValueWriters.map(key, value);
     }
 
     @Override
-    public void addValue(int rowId, int column, SpecializedGetters data,
-                         ColumnVector output) {
-      if (data.isNullAt(column)) {
-        output.noNulls = false;
-        output.isNull[rowId] = true;
-      } else {
-        output.isNull[rowId] = false;
-        ArrayData value = data.getArray(column);
-        ListColumnVector cv = (ListColumnVector) output;
-        // record the length and start of the list elements
-        cv.lengths[rowId] = value.numElements();
-        cv.offsets[rowId] = cv.childCount;
-        cv.childCount += cv.lengths[rowId];
-        // make sure the child is big enough
-        cv.child.ensureSize(cv.childCount, true);
-        // Add each element
-        for (int e = 0; e < cv.lengths[rowId]; ++e) {
-          children.addValue((int) (e + cv.offsets[rowId]), e, value, cv.child);
-        }
-      }
-    }
-  }
-
-  static class MapConverter implements Converter {
-    private final Converter keyConverter;
-    private final Converter valueConverter;
-
-    MapConverter(TypeDescription schema) {
-      keyConverter = buildConverter(schema.getChildren().get(0));
-      valueConverter = buildConverter(schema.getChildren().get(1));
+    public SparkOrcValueWriter primitive(Type.PrimitiveType iPrimitive, TypeDescription primitive) {
+      switch (primitive.getCategory()) {
+        case BOOLEAN:
+          return SparkOrcValueWriters.booleans();
+        case BYTE:
+          return SparkOrcValueWriters.bytes();
+        case SHORT:
+          return SparkOrcValueWriters.shorts();
+        case DATE:
+        case INT:
+          return SparkOrcValueWriters.ints();
+        case LONG:
+          return SparkOrcValueWriters.longs();
+        case FLOAT:
+          return SparkOrcValueWriters.floats();
+        case DOUBLE:
+          return SparkOrcValueWriters.doubles();
+        case BINARY:
+          return SparkOrcValueWriters.byteArrays();
+        case STRING:
+        case CHAR:
+        case VARCHAR:
+          return SparkOrcValueWriters.strings();
+        case DECIMAL:
+          return SparkOrcValueWriters.decimal(primitive.getPrecision(), primitive.getScale());
+        case TIMESTAMP_INSTANT:
```

Comment:
I think we need to take into account if we need to adjust for UTC. This information is present in `iPrimitive`.  I think we should dispatch on the `iprimitive type` instead of the ORC primitive type, no?

Reply:
I'm not quite sure what you mean here.

Iceberg must never modify a value that is passed in, other than to convert to a different representation. Adjusting for time zones would produce corrupt values.

It sounds like you're talking about the `adjust-to-utc` property that we set for Avro and Parquet, which tells us whether a timestamp is `with time zone` (adjust=true) or `without time zone` (adjust=false). But those only determine the type. Values are never modified because of that type. In this case, we don't need to do that because TIMESTAMP_INSTANT is used for timestamp with time zone, and TIMESTAMP is used for timestamp without time zone. So in this branch of the switch, the type is definitely timestamptz.

Reply:
> So in this branch of the switch, the type is definitely timestamptz.

Yes,  in this `TIMESAMP_INSTANT` case, it must be a timestamp with time zone. see the Category in TypeDescription: 

```java
public enum Category {
    BOOLEAN("boolean", true),
    BYTE("tinyint", true),
    SHORT("smallint", true),
    INT("int", true),
    LONG("bigint", true),
    FLOAT("float", true),
    DOUBLE("double", true),
    STRING("string", true),
    DATE("date", true),
    TIMESTAMP("timestamp", true),
    BINARY("binary", true),
    DECIMAL("decimal", true),
    VARCHAR("varchar", true),
    CHAR("char", true),
    LIST("array", false),
    MAP("map", false),
    STRUCT("struct", false),
    UNION("uniontype", false),
    TIMESTAMP_INSTANT("timestamp with local time zone", false);
```

btw, Is there any reason that we spark don't support a timestamp without time zone ?   I saw the spark test [here](https://github.com/apache/iceberg/blob/ffdcf09027e09460b7d7505e65aea119107934a3/spark/src/test/java/org/apache/iceberg/spark/data/AvroDataTest.java#L52) did not include the `timestamp without zone` data type.  Maybe we could file another issue to address this `timestamp without zone` issue, because this is a refactor PR and we could get this merged firstly.

Reply:
I was thinking that we should also handle `TIMESTAMP` and not just `TIMESTAMP_INSTANT` . But I checked with @shardulm94  and the reason is that Spark does not support timestamp without timezone.

My other point was to dispatch on primitive so that it is consistent with other writers. e.g GenericOrcWriter.  But here it doesn't make a lot of difference.

Thanks for the explanation!